File: src/main/java/wraith/fabricaeexnihilo/json/recipe/LeakingRecipeJson.java
package wraith.fabricaeexnihilo.json.recipe;

import com.google.gson.*;
import wraith.fabricaeexnihilo.api.crafting.FluidIngredient;
import wraith.fabricaeexnihilo.api.crafting.ItemIngredient;
import wraith.fabricaeexnihilo.api.recipes.barrel.LeakingRecipe;
import wraith.fabricaeexnihilo.json.BaseJson;
import wraith.fabricaeexnihilo.json.basic.BlockJson;

import java.lang.reflect.Type;

public final class LeakingRecipeJson extends BaseJson<LeakingRecipe> {

    private LeakingRecipeJson() {}
    
    public static final LeakingRecipeJson INSTANCE = new LeakingRecipeJson();

    @Override
    public LeakingRecipe deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        var obj = json.getAsJsonObject();
        return new LeakingRecipe(
                ItemIngredient.fromJson(obj.get("target"), context),
                FluidIngredient.fromJson(obj.get("fluid"), context),
                context.deserialize(obj.get("loss"), INT_TYPE_TOKEN),
                BlockJson.INSTANCE.deserialize(obj.get("result"), BlockJson.INSTANCE.getTypeToken(), context)
        );
    }

    @Override
    public JsonElement serialize(LeakingRecipe src, Type typeOfSrc, JsonSerializationContext context) {
        var obj = new JsonObject();
        obj.add("target", src.target().toJson(context));
        obj.add("fluid", src.fluid().toJson(context));
        obj.add("loss", context.serialize(src.loss()));
        obj.add("result", BlockJson.INSTANCE.serialize(src.result(), BlockJson.INSTANCE.getTypeToken(), context));
        return obj;
    }

}

File: src/main/kotlin/exnihilofabrico/json/LeakingRecipeJson.kt
package exnihilofabrico.json

import com.google.gson.*
import com.google.gson.reflect.TypeToken
import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.api.crafting.ItemIngredient
import exnihilofabrico.api.recipes.barrel.LeakingRecipe
import java.lang.reflect.Type

object LeakingRecipeJson: JsonDeserializer<LeakingRecipe>, JsonSerializer<LeakingRecipe> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): LeakingRecipe {
        val obj = json.asJsonObject
        return LeakingRecipe(
            ItemIngredient.fromJson(obj["target"], context),
            FluidIngredient.fromJson(obj["fluid"], context),
            context.deserialize(obj["loss"], object : TypeToken<Int>() {}.type),
            BlockJson.deserialize(obj["result"], BLOCK_TYPE_TOKEN, context)
        )
    }

    override fun serialize(src: LeakingRecipe, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        obj.add("target", src.target.toJson(context))
        obj.add("fluid", src.fluid.toJson(context))
        obj.add("loss", context.serialize(src.loss))
        obj.add("result", BlockJson.serialize(src.result, BLOCK_TYPE_TOKEN, context))
        return obj
    }

}